src/rsa/blinding.rs

```diff
@@ -0,0 +1,644 @@
+    const SUCCESS: c::int = 1;
+
+    let max_exclusive = core::slice::from_raw_parts(max_exclusive, num_limbs);
+    let mut dest = core::slice::from_raw_parts_mut(dest, num_limbs);
+
+    let result = set_to_rand_mod(&mut dest, &max_exclusive, (*rng).rng);
+    if result.is_err() {
+        return ERR;
+    }
+
+    SUCCESS
+}
+
+/// Chooses a positive integer less than `max_exclusive` uniformly at random
+/// and stores it into `dest`.
+fn set_to_rand_mod(dest: &mut [Limb], max_exclusive: &[Limb], rng: &rand::SecureRandom)
+            -> Result<(), error::Unspecified> {
+
+    debug_assert_eq!(dest.len(), max_exclusive.len());
+
+    let sampling_params = select_sampling_params(max_exclusive);
+
+    if sampling_params.extend_limbs_by_one {
+        with_extended_buffers(dest, max_exclusive, |ext_dest, ext_max| {
+            let range = Range {
+                max_exclusive: ext_max,
+                sampling_params: &sampling_params,
+            };
+            range.sample_into_limbs(ext_dest, rng)
+        })
+    } else {
+        let range = Range {
+            max_exclusive: max_exclusive,
+            sampling_params: &sampling_params,
+        };
+        range.sample_into_limbs(dest, rng)
+    }
+}
+
+/// Copy `dest` and `max_exclusive` into temporary buffers which extend capacity
+/// by one, pass those to the callback, then copy the modified destination
+/// buffer back to `dest`.
+fn with_extended_buffers<F>(dest: &mut [Limb], max_exclusive: &[Limb], cb: F)
+                            -> Result<(), error::Unspecified>
+                            where F: Fn(&mut [Limb], &[Limb])
+                                     -> Result<(), error::Unspecified> {
+    const BUF_SIZE: usize = RSA_KEY_MAX_LIMBS + 1;
+
+    let buf_needed = dest.len() + 1;
+    debug_assert_eq!(buf_needed, max_exclusive.len() + 1); // equivalent definition
+
+    assert!(buf_needed <= BUF_SIZE);
+
+    let mut tmp_max: [Limb; BUF_SIZE] = [0; BUF_SIZE];
+    let mut tmp_dest: [Limb; BUF_SIZE] = [0; BUF_SIZE];
+
+    tmp_dest[..dest.len()].clone_from_slice(&dest);
+    tmp_max[..max_exclusive.len()].clone_from_slice(&max_exclusive);
+
+    let result = cb(&mut tmp_dest[..buf_needed], &tmp_max[..buf_needed]);
+
+    if result.is_ok() {
+        let dest_len = dest.len();
+        dest.clone_from_slice(&tmp_dest[..dest_len]);
+    }
+
+    result
+}
+
+/// Decide implementation strategy for random sampling - see `SamplingParams`.
+fn select_sampling_params(max_exclusive: &[Limb]) -> SamplingParams {
+    let starts_with_0b100 = starts_with_0b100_variable_time(max_exclusive);
+    let leading_zeros = max_exclusive[max_exclusive.len() - 1].leading_zeros();
+
+    if starts_with_0b100 && leading_zeros == 0 {
+        SamplingParams {
+            most_sig_limb_mask: 1, // effectively a carry into a new, more-significant limb
+            reduce_when_over_bound: true,
+            extend_limbs_by_one: true,
+        }
+    } else if starts_with_0b100 {
```

Comment:
I feel like we don't need to treat this `else if` case differently than the `else` case below. It is very unlikely that the modulus isn't going to be a multiple of 64-bits in any case where the application cares about performance. This means we can combine `reduce_when_over_bound` and `extend_limbs_by_one` into a single variable.